Nexus 3 metadata, npm folder and checksum file issue solutions

Shop-kins opened the pull request:
our systems is a Nexus 3 instance initial migrated over from a nexus 2 instance. This resulted in us maintaining a system of storing md5 and sha1 files alongside our artifacts. When using this migrator it picked up these files and attempted to upload them to Artifactory, not only is this a fruitless task as artifactory manages that for us, when it attempts the `checksum upload` the checksum files contents is not included, and as Artifactory is a clever thing it knows its a checksum and looks inside the file, resulting in some empty string issues.

Not only this but we discovered intermittent issues with npm repos being uploaded, this oroginated from the folders metadata being uploaded as a file and thus blocking any further files in that directory structure.

We also have a reliance on timestamps in certain areas, and this migration process replaces them with the uploaded time, there is a change in here which alters this.

And these changes resulted in a successful migration

ShopAnaplan committed:
Ensure blob to be uploaded is not a checksum file
This avoids a lot of empty string checksum issues caused by a quirk with `checksum uploading` a checksum

ShopAnaplan committed:
Ensure blob to be uploaded is not an npm folder
This is to ensure that no npm folder metadata files get created, as these block the folders contents being uploaded

ShopAnaplan committed:
Ensure metadata for nexus3 artifacts is pulled across
Timestamps for the artifacts are incredibly important for some of our teams workflows, and so the original nexus one is required

DarthFennec commented on nex2art/core/Upload.py:
```diff
@@ -196,11 +199,15 @@ def runThread(self, queue, url, headers):
             else: store = self.acquireLocation2(path, metapath)
             paths = self.deployPaths(path, metapath, repo, store)
             for lpath, mpath, rep, rpath, props in paths:
+                localheaders = headers
```
I feel like what you mean to do here is make a copy of `headers` so the original doesn't get modified when you change `localheaders`. That's not what this does. I think this should be `localheaders = headers.copy()`

Shop-kins replied:
good catch!

DarthFennec commented on nex2art/core/Upload.py:
```diff
@@ -398,3 +405,13 @@ def incFileCount(self, fname, error=False):
+    def isNexus3ChecksumFile(self, chapdir, fname):
+        blobpropertiesfile = os.path.join(chapdir, fname + '.properties')
+        blobproperties = open(blobpropertiesfile, 'r').read()
+        return ('.md5' in blobproperties) or ('.sha1' in blobproperties) or ('.sha256' in blobproperties)
```
Is this the best way to do it? We don't want to skip artifacts that just happen to have one of these strings somewhere in their metadata. Maybe instead something like:
``` python
blobname = self.acquireMetadata3(blobpropertiesfile)['@BlobStore.blob-name']
return blobname.endswith('.md5') or blobname.endswith('.sha1') or blobname.endswith('.sha256')
```

Shop-kins replied:
I like this, a much more conservative solution

DarthFennec commented on nex2art/core/Upload.py:
```diff
@@ -398,3 +405,13 @@ def incFileCount(self, fname, error=False):
+    def isNexus3NpmFolder(self, chapdir, fname):
+        blobpropertiesfile = os.path.join(chapdir, fname + '.properties')
+        blobproperties = open(blobpropertiesfile, 'r').read()
+        return (re.search('blob-name=@', blobproperties)) and (not re.search('blob-name=@.+\..*', blobproperties))
```
This is actually already dealt with, in [this file](https://github.com/jfrog/nexus2artifactory/blob/master/nex2art/core/Npm.py). If you're getting this bug, the logic in that file might be too narrow, in which case you should probably modify it there instead of adding this here.

Shop-kins replied:
After looking through this I've located the part which our folders are sneaking through. The self.checkContent(), as all the name properties in the blob bytes files are the full `@scope/folder` not just the `folder`. I've altered this to be `js['name'].endswith(name)` from `js['name'] == name` in an attempt to retain any existing cases covered by this.

DarthFennec reviewed (changes requested):
I really like this. I've considered preserving migrated timestamps, but I didn't think it was actually possible to do, so it's fantastic to see that this works for you. As for checksum files, those are ignored by the Nexus 2 migration code, but I guess the Nexus 3 code doesn't bother since those files are no longer standard. You're right in that they should be ignored, and that they would exist in a system migrated from Nexus 2.

The NPM metadata folder issue was discovered during early testing, and the migrator does already account for it. If you're encountering this issue, there may be a bug in the existing handler, then. I expect the check I'm making is less broad than it should be.

Thanks for the contribution!

ShopAnaplan committed:
More appropriate NPM fix, and conservative checksum change
To ensure that the checksum change only covers the case we desire this has been altered to look only at blob-name, the npm change is an attempt to cover previous cases as well as the new.

Shop-kins commented:
I've commented on the requests individually, the safe checksum and copy headers seem solid. after looking through the npm file and some more digging through our blob files, the case I'm trying to cover is when the blobs bytes file has its name as `@scope/folder` and the code attempts to compare it to simply `folder`. The small change i've made attempts to account for this.

DarthFennec commented:
Awesome, looks good to me. Merged.
